dask/core.py

```diff
@@ -404,11 +404,33 @@ def _toposort(dsk, keys=None, returncycle=False, dependencies=None):
                 if nxt not in completed:
                     if nxt in seen:
                         # Cycle detected!
-                        cycle = [nxt]
+                        # Let's report only the nodes that directly participate in the cycle.
+                        # We use `priorities` below to greedily construct a short cycle.
+                        # Shorter cycles may exist.
+                        priorities = {}
+                        prev = nodes[-1]
+                        # Give priority to nodes that were seen earlier.
                         while nodes[-1] != nxt:
-                            cycle.append(nodes.pop())
-                        cycle.append(nodes.pop())
+                            priorities[nodes.pop()] = -len(priorities)
```

Comment:
I do not understand `priorities` here. Hash in Python 3 random -> dependency_set random order -> stack random order -> priorities random, so we are just picking a random `prev dep`?

Reply:
`priorities` is used as the key in the `min` call below to construct shorter chains.  This let's us take bigger steps when there are multiple options when "walking backwards", and bigger steps results in fewer steps (hopefully).

To clarify, we use `priorities.__getitem__` as the key function for `min`.  For example, `key={1: 10, 2:20}.__getitem__` is a function that maps `1` to `10` and `2` to `20`.